i18n the AccessDeniedException message

dkodippily opened the pull request:
Fixes:#9823

<!--
For Security Vulnerabilities, please use https://pivotal.io/security#reporting
-->

<!--
Before creating new features, we recommend creating an issue to discuss the feature. This ensures that everyone is on the same page before extensive work is done.

Thanks for contributing to Spring Security. Please provide a brief description of your pull-request and reference any related issue numbers (prefix references with gh-).
-->

sjohnr reviewed (commented):
Hi @dkodippily, see my feedback below. We'll spend some time iterating and discussing to see what we come up with, if that's ok with you.

dkodippily commented on core/src/main/java/org/springframework/security/access/AccessDeniedException.java:
```diff
@@ -24,6 +27,8 @@
  */
 public class AccessDeniedException extends RuntimeException {
 
+	protected final MessageSourceAccessor messages = SpringSecurityMessageSource.getAccessor();
```
@sjohnr Thanks for the feedback, working on these changes. Do you have any thoughts on the below class naming and usage.
`class MessageSourceAccessProvider {

    private MessageSourceAccessProvider(){};

    private static class MessageSourceAccessorHolder {
        static final MessageSourceAccessor INSTANCE = SpringSecurityMessageSource.getAccessor();
    }

    public static MessageSourceAccessor getInstance() {
        return MessageSourceAccessorHolder.INSTANCE;
    }
}`

sjohnr replied:
The class looks good. Since we're just focusing on access denied right now, how about `AccessDeniedMessageSourceHolder`? (`Provider` is nice too, so it could work.)

Might want to make the `getInstance()` package-private as well. Not sure it matters, but seems best for consistency.

dkodippily replied:
@sjohnr indeed AccessDeniedMessageSourceHolder make more sense.

sjohnr commented on core/src/test/java/org/springframework/security/access/AccessDeniedExceptionTests.java:
```diff
@@ -0,0 +1,75 @@
+		Locale beforeHolder = LocaleContextHolder.getLocale();
+		LocaleContextHolder.setLocale(Locale.FRANCE);
+
+		org.springframework.security.access.AccessDeniedException accessDeniedException = new org.springframework.security.access.AccessDeniedException(
```
I believe you can safely add an import for `AccessDeniedException` to replace fully qualified references.

sjohnr commented on core/src/main/java/org/springframework/security/access/AccessDeniedMessageSourceHolder.java:
```diff
@@ -0,0 +1,42 @@
+final class AccessDeniedMessageSourceHolder {
+
+	private AccessDeniedMessageSourceHolder() {
+	};
```
Remove semicolon.

sjohnr commented on core/src/main/java/org/springframework/security/access/AccessDeniedMessageSourceHolder.java:
```diff
@@ -0,0 +1,42 @@
+/**
+ * Provide a {@link org.springframework.context.support.MessageSourceAccessor} singleton
+ *
+ * @author Dayan Kodippily - 23/7/21
```
Remove ` - 23/7/21` and add a new tag with `@since 5.6`. See [Use @since tags](https://github.com/spring-projects/spring-security/blob/main/CONTRIBUTING.adoc#use-since-tags-for-newly-added-public-api-types-and-methods).

sjohnr commented on core/src/main/java/org/springframework/security/access/AccessDeniedException.java:
```diff
@@ -42,4 +42,12 @@ public AccessDeniedException(String msg, Throwable cause) {
 	}
 
+	/**
+	 * @return The localized description of this <code>AccessDeniedException</code>.
```
I would suggest adding a javadoc message before the `@return`, such as `Returns the localized description of this {@code AccessDeniedException}, by making use of a {@link SpringSecurityMessageSource} using the exception message as the key.`

You should also change `<code>` in `@return` to `{@code AccessDeniedException}` and remove the period at the end. Most returns messages start with a lower-case letter, so you may want to change "The" to "the" as well.

sjohnr commented on core/src/test/java/org/springframework/security/access/AccessDeniedExceptionTests.java:
```diff
@@ -0,0 +1,75 @@
+		Locale beforeSystem = Locale.getDefault();
+		Locale.setDefault(Locale.US);
+		Locale beforeHolder = LocaleContextHolder.getLocale();
+		LocaleContextHolder.setLocale(Locale.US);
```
Can these steps and the steps at the end of the method be incorporated into an `@BeforeEach` and `@AfterEach`?

dkodippily replied:
These changes are added

sjohnr commented on core/src/main/java/org/springframework/security/access/AccessDeniedMessageSourceHolder.java:
```diff
@@ -0,0 +1,42 @@
+
+	private static class MessageSourceAccessorHolder {
+
+		static final MessageSourceAccessor INSTANCE = SpringSecurityMessageSource.getAccessor();
```
Another minor suggestion, but here I think you're safe with `private` to guarantee even the package-level doesn't access it directly.

sjohnr reviewed (changes requested):
Hi @dkodippily, thanks for the updates. Please see some additional and more specific feedback inline.

Also please review [contributor guidelines](https://github.com/spring-projects/spring-security/blob/main/CONTRIBUTING.adoc). In particular, please squash your commits and modify the message to include "Closes gh-9823".

dkodippily commented:
> Hi @dkodippily, thanks for the updates. Please see some additional and more specific feedback inline.
> 
> Also please review [contributor guidelines](https://github.com/spring-projects/spring-security/blob/main/CONTRIBUTING.adoc). In particular, please squash your commits and modify the message to include "Closes [gh-9823](https://github.com/spring-projects/spring-security/issues/9823)".

dkodippily commented:
@sjohnr I fixed all the changes but accidentally closed it

sjohnr reviewed (commented):
Hello @dkodippily, thanks so much for the changes. This is great so far, and hopefully it's giving you good experience in making these contributions! Please see the below comment. I'm interested in your thoughts at this point, since you've put personal time into this.

Lastly, just a reminder that commit messages should be of the form:

```
i18n the AccessDeniedException message

Closes gh-9823
```

dkodippily commented on core/src/main/java/org/springframework/security/access/AccessDeniedException.java:
```diff
@@ -42,4 +42,15 @@ public AccessDeniedException(String msg, Throwable cause) {
+	 */
+	public String getLocalizedMessage() {
+		String message = this.getMessage();
+		return AccessDeniedMessageSourceHolder.getInstance().getMessage(message, message);
```
Thanks for the update, yes I'm really interested in continuing. I'll also dig deep into this and push the changes asap.

sjohnr replied:
Ok, great. I think for discussing option 2 we can go back to the original issue and comment there.

dkodippily replied:
with regards to point 1 and 2 we can confirm if the it manages to find/miss a message with the current local by calling the single argument getMessage constructor and handling the NoSuchMessageException. Then we call the getMessage with original message again as the default value and US Local. any thoughts on below implementation?

`	public String getLocalizedMessage() {
		String message = this.getMessage();
		String translatedMessage = null;
		try {
			translatedMessage = AccessDeniedMessageSourceHolder.getInstance().getMessage(message);
		}catch(NoSuchMessageException ignore){

		}

		if(translatedMessage != null){
			return translatedMessage;
		}

		return AccessDeniedMessageSourceHolder.getInstance().getMessage(message, message, Locale.US);
	}`

dkodippily replied:
@sjohnr , pushed all the changes.

dkodippily commented:
Hi @sjohnr , any updates on this ?

sjohnr commented:
Hey @dkodippily, sorry for the radio silence. I've had a number of days off recently, and I'm currently heads down preparing for SpringOne, so haven't had much time. Regarding your last comment, that is similar to what I was thinking.

However, I recommend you head back over to the original issue and see about a different proposal for solving the issue, because the path I was hoping for didn't pan out. Specifically, I'm saying we should seek a solution that doesn't modify `AccessDeniedException`. The reason is because this solution requires numerous changes throughout the framework to use it, and still could break passivity for consumers of the framework that handle an `AccessDeniedException`.

Hopefully that makes sense, but let me know if it's still unclear.

sjohnr commented:
Just a heads up that I'm closing this P/R (see the above comment). Thanks so much for working through the experiment and addressing all the comments (good practice run for other PRs)! We'll pick up the conversation again with a different approach in #9823.

dkodippily commented:
> Just a heads up that I'm closing this P/R (see the above comment). Thanks so much for working through the experiment and addressing all the comments (good practice run for other PRs)! We'll pick up the conversation again with a different approach in #9823.

@sjohnr Thanks for the update, sorry for the long pause without any updates, crazy few weeks. I was thinking, if we are not going to touch the AccessDeniedException class how about a subclass with i18n support, also I'll check other i18n aware interfaces as originally mentioned in the issue. Thanks for insights, was a great learning.

sjohnr commented:
@dkodippily, thinking out loud I'm not sure whether a subclass would be a good way to go or not, but happy to discuss further. Thanks!
